examples/step-95/step-95.cc

```diff
@@ -0,0 +1,1711 @@
+    template <bool is_dg_>
+    void do_local_apply_gp_face_term(FaceIntegrator            &evaluator_m,
+                                     FaceIntegrator            &evaluator_p,
+                                     const VectorizedArrayType *dof_ptr_m,
+                                     const VectorizedArrayType *dof_ptr_p,
+                                     const VectorizedArrayType &diameter,
+                                     const bool sum_into_values) const
+    {
+      EvaluationFlags::EvaluationFlags evaluation_flags =
+        EvaluationFlags::gradients;
+      if (is_dg_)
+        evaluation_flags |= EvaluationFlags::values;
+
+      const unsigned int degree =
+        matrix_free->get_dof_handler(dof_index).get_fe().degree;
+      const bool do_hessians = degree > 1;
+      if (do_hessians)
+        evaluation_flags |= EvaluationFlags::hessians;
+
+      Assert(degree <= 2,
+             ExcMessage(
+               "Face-based stabilization only implemented up to degree 2!"));
+
+      evaluator_m.evaluate(dof_ptr_m, evaluation_flags);
+      evaluator_p.evaluate(dof_ptr_p, evaluation_flags);
+
+      const VectorizedArrayType factor_values   = 0.5 / diameter;
+      const VectorizedArrayType factor_gradient = 0.5 * diameter;
+      const VectorizedArrayType factor_hessians =
+        0.5 * diameter * diameter * diameter;
+
+      if (do_hessians)
+        for (const auto q : evaluator_m.quadrature_point_indices())
+          do_gp_face_term<true, is_dg_>(evaluator_m,
+                                        evaluator_p,
+                                        factor_values,
+                                        factor_gradient,
+                                        factor_hessians,
+                                        q);
+      else
+        for (const auto q : evaluator_m.quadrature_point_indices())
+          do_gp_face_term<false, is_dg_>(evaluator_m,
+                                         evaluator_p,
+                                         factor_values,
+                                         factor_gradient,
+                                         factor_hessians,
+                                         q);
+
+      evaluator_m.integrate(evaluation_flags, sum_into_values);
+      evaluator_p.integrate(evaluation_flags, sum_into_values);
+    }
+
+    // Three helper functions to determine the category of face based on the
+    // active_fe_index of the face-sharing cells.
+    inline bool
+    is_inside_face(std::pair<unsigned int, unsigned int> face_category) const
+    {
+      return is_inside(face_category.first) && is_inside(face_category.second);
+    }
+
+    inline bool
+    is_mixed_face(std::pair<unsigned int, unsigned int> face_category) const
+    {
+      return (is_inside(face_category.first) &&
+              is_intersected(face_category.second)) ||
+             (is_intersected(face_category.first) &&
+              is_inside(face_category.second));
```

Comment:
Could it make sense to state
```suggestion
      return !(is_inside_face(face_category) ||
              is_intersected_face(face_category.second));
```

Reply:
I think we also have 'outside' faces (between `FE_Nothing` elements). I would keep this as is.